flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoRequestCoordinator.java

```diff
@@ -0,0 +1,188 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.webmonitor.threadinfo;
+
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.AccessExecution;
+import org.apache.flink.runtime.executiongraph.AccessExecutionVertex;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.messages.TaskThreadInfoResponse;
+import org.apache.flink.runtime.messages.ThreadInfoSample;
+import org.apache.flink.runtime.taskexecutor.TaskExecutorGateway;
+import org.apache.flink.runtime.webmonitor.stats.TaskStatsRequestCoordinator;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** A coordinator for triggering and collecting thread info stats of running operator subtasks. */
+public class ThreadInfoRequestCoordinator
+        extends TaskStatsRequestCoordinator<List<ThreadInfoSample>, OperatorThreadInfoStats> {
+
+    /**
+     * Creates a new coordinator for the job.
+     *
+     * @param executor Used to execute the futures.
+     * @param requestTimeout Time out after the expected sampling duration. This is added to the
+     *     expected duration of a request, which is determined by the number of samples and the
+     *     delay between each sample.
+     */
+    public ThreadInfoRequestCoordinator(Executor executor, long requestTimeout) {
+        super(executor, requestTimeout);
+    }
+
+    /**
+     * Triggers collection of thread info stats of an operator by combining thread info responses
+     * from given subtasks. A thread info response of a subtask in turn consists of {@code
+     * numSamples}, collected with {@code delayBetweenSamples} milliseconds delay between them.
+     *
+     * @param subtasksWithGateways Execution vertices together with TaskExecutors running them.
+     * @param numSamples Number of thread info samples to collect from each subtask.
+     * @param delayBetweenSamples Delay between consecutive samples (ms).
+     * @param maxStackTraceDepth Maximum depth of the stack traces collected within thread info
+     *     samples.
+     * @return A future of the completed thread info stats.
+     */
+    public CompletableFuture<OperatorThreadInfoStats> triggerThreadInfoRequest(
+            List<Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>>
```

Comment:
This parameter type is inducing nightmares ;) Maybe add a small record class?

Reply:
This code would be easier to test if it wouldn't specifically require a TaskExecutorGateway; we could move the newly added method into a separate interface that the TM extends.